py/objtype.c

```diff
@@ -852,7 +852,7 @@ STATIC mp_obj_t instance_subscr(mp_obj_t self_in, mp_obj_t index, mp_obj_t value
         meth_args = 3;
     }
     if (member[0] == MP_OBJ_SENTINEL) {
-        return mp_obj_subscr(self->subobj[0], index, value);
+        return mp_obj_subscr_impl(self->subobj[0], index, value, instance);
```

Comment:
Did you try passing `self` here instead of `self->subobj[0]`? You'll need to add it to all of the native calls where you modified method signature. That will ensure that all native subscrs receive the actual object (native or not).

Reply:
I had planned to do this - now done.

Reply:
These changes aren't what I meant. I meant leaving the signature as it was originally and passing self here rather than the native "subobject". To do so, each native subscr will have to get the native object itself instead. (Except for yours which will use both.)

Reply:
I'll have a look around and see if that approach would work.  I feel like I tried that and ended up in an infinite loop because the calling path is:

mp_obj_subscr => instance_subscr => mp_obj_subscr
(because of o->subscr = instance_subscr in mp_obj_new_type)

Reply:
Ah right! Seems to me that instance_subscr should be smarter itself rather than calling back to mp_obj_subscr.

Reply:
If you can think of how to make instance_subscr smarter, i'd love to know your thoughts.  Ultimately it does that when the base is native and has a subscr (see mp_obj_class_lookup ~ line 168) `lookup->dest[0] = MP_OBJ_SENTINEL;`

I guess we could make it look up and call the native subscr method itself, but then all of the assertions mp_obj_subscr could throw need reimplementing.

Reply:
So, I think the native call needs to be moved from mp_obj_subscr to obj_instance. As is, a subclass of a native object won't be able to override subscr. This won't require reimplementing or moving the errors thrown by mp_obj_subscr because instance_subscr can return MP_OBJ_NULL to prevent the early return and cause an error.

The native call also needs to pass in the original instance and each native implementation will need to get the subobj to use. That will allow the native subscr full access to the instance.

Reply:
If I understand you correctly, instead of instance_subscr calling mp_obj_subscr, it should do something like this:

```
    if (member[0] == MP_OBJ_SENTINEL) {
        //return mp_obj_subscr(self->subobj[0], index, value, instance);

        mp_obj_type_t *type = mp_obj_get_type(self->subobj[0]);
        if (type->subscr != NULL) {
            mp_obj_t ret = type->subscr(self_in, index, value, instance);
            // May have called port specific C code. Make sure it didn't mess up the heap.
            assert_heap_ok();
            if (ret != MP_OBJ_NULL) {
                return ret;
            }
        }
        return MP_OBJ_NULL;
```

This fails the unit tests.  Much like many of my first attempts to make this work.

> 2 tests failed: class_store_class subclass_native1

### Expected
b'[1, 2, 5]\nsomething\n5\n[-1, 2, 5]\n3\n[-1, 2, 5, 20, 30, 40]\nTypeError\nTypeError\n'
### Actual
b'[1, 2, 5]\nsomething\nsomething\n[1, 2, 5]\n3\n[1, 2, 5, 20, 30, 40]\nTypeError\nTypeError\n'
FAIL  basics/subclass_native1.py
### Expected
b"b'a' b'b'\nAll's ok\n"
### Actual
b'CRASH'
FAIL  basics/class_store_class.py
2 tests performed (10 individual testcases)
0 tests passed
2 tests failed: class_store_class subclass_native1

Reply:
Did you update list's subscr as well? It'll need to get subobj[0] itself. Not doing so could explain the first failure.

The second will fail if namedtuple's native object subscr hasn't been updated.

Reply:
I did not update any subscr methods.  I'll give those two a try, and if that works, I'll know what to check for codebase-wide.

Reply:
@tannewt while working on this approach, I was partly successful. 
In instance_subscr i fetched `subobj[0]` to get the type, then called `subobj_type->subscr(self_in, index, value).`

For some of the _subscr implementations I used `mp_instance_cast_to_native_base`, for example, in `list_subscr` which appeared to work, and satisfied the test suite... But when it came to `bytes_subscr` which is used by both `str` and `bytes`, I couldn't find a straightforward solution that didn't break the test suite (or crash).

Reply:
Getting the subobj directly in bytes_subscr didn't work? Wouldn't that be the equivalent of what it's doing now?

Reply:
I ended up with crashes.  I'm wary of significantly modifying the code here without fully understanding it.  
https://github.com/adafruit/circuitpython/blob/master/py/objstr.c#L443

Reply:
Is the change anywhere? I'd be happy to poke at it. Was it tests that were crashing?

Reply:
@tannewt https://github.com/rhooper/circuitpython/tree/pixelbuf-subscr-change is the WIP

Reply:
So the crash thing is not as bad as it sounds. The class_store_class test was just throwing an exception that subscr wasn't supported. I've fixed it here: https://github.com/rhooper/circuitpython/pull/2

I don't think there is much you need to change actually. Most of the native subscr methods are called directly rather than through instance_subscr. instance_subscr only comes into play with native class when they are subclassed. Most native objects don't work correctly when subclassed so that isn't generally an issue. namedtuples are a very special case where they are subclassable and they themselves pretend to be a subclass of tuple.